k2/python/k2/fsa.py

```diff
@@ -26,6 +26,57 @@ class Fsa(object):
     a vector of FSAs it is a tuple with three
     elements ``(num_fsas, None, None)``.  (Caution: it's possible
     for a vector of FSAs to have zero or one elements).
+
+    An instance of FSA has the following attributes:
+
+    - ``arcs``: You will NOT use it directly in Python. It is an instance
+                of ``_k2.RaggedArc`` with only one method ``values()`` which
+                returns a 2-D `torch.Tensor`` of dtype ``torch.int32`` with 4
+                columns. Its number of rows indicates the number of arcs in the
+                FSA. The first column represents the source states, second
+                column the destination states, third column the labels and the
+                fourth column is the score. Note that the score is actually
+                a float number but it is **reinterpreted** as an integer.
+
+    - ``score``: A 1-D ``torch.Tensor`` of dtype ``torch.float32``. It has
+                 as many entries as the number of arcs representing the score
+                 of every arc.
+
+    - ``labels``: A 1-D ``torch.Tensor`` of dtype ``torch.int32``. It has as
+                  many entries as the number of arcs representing the label of
+                  every arc.
+
+    - ``arc_labels_sorted``: A boolean. True if arcs of every state are sorted
+                             by labels in increasing order. False means arcs
+                             are not sorted by ``labels``.
+
+    It MAY have the following attributes:
+
+    - ``symbols``: An instance of ``k2.SymbolTable``. It maps an entry in
+                   ``labels`` to an integer and vice versa. It is used for
+                   visualization only.
+
+    - ``aux_labels`: A 1-D ``torch.Tensor`` of dtype ``torch.int32``. It has the
+                     same shape as ``labels``.
+
+    - ``aux_symbols``: An instance of ``k2.SymbolTable. It maps an entry in
+                       ``aux_labels`` to an integer and vice versa.
+
+    - ``arc_aux_labels_sorted``: A boolean. True if arcs of every state are
```

Comment:
Actually it seems to me a little strange that arcs sorted by `axu_labels`, as I was thinking we just sort arcs on labels (though they may be inverted)

Reply:
If an `fsa` is already arc sorted. Do you have a better method to determine that `fsa.invert_().invert_()` is also arc sorted?

Reply:
The most direct way, IMO, to tell whether things are arc-sorted, is the properties.  
There is a flag for that when we call Properties().
One of the things the Fsa object can cache is the 'properties' tensor from GetFsaBasicProperties() (one property per FSA) and the 'properties' scalar which is the "and" of the properties tensor.

It might be simplest just to recompute the properties when we invert (?), because I don't think inversion will happen veyr often.

Reply:
I mean, if we just need a value to cache the property of `arc_sorted`, why not just use a local variable to do this? Exposing this    at the Fsa interface level is confusing as we never really sort an Fsa on aux_lables (what we do is:  invert then sort)

Reply:
For `c = k2.compose(a, b)`, we need to ensure that both `a.aux_labels`  and `b.labels` are sorted. I'm afraid it's difficult to know whether an `fsa` is arc sorted with a single variable `arc_sorted`.

The intention of these two variables is to avoid calling `arc_sort`  when an `fsa` is already arc sorted.
If `k2.compose()` is called `N` times, then `invert_()` will be invoked `3N` times. I am not sure how often `k2.compose` will be invoked.

> One of the things the Fsa object can cache

I think we have to figure out which properties are changed by which operations and update the cache accordingly.

Reply:
> what we do is: invert then sort)

`arc_aux_labels_sorted` is to avoid `arc_sort` after calling `invert_()`.

Reply:
mm, I mean, can't we just cache `arc_sorted` in a local variable and then, after `a.invert().invert()`, set the local variable to the result fsa's `arc_sorted`. Or just as Dan said, check the properits again after `invert().invert()`, it's not so expensive considering we don't call `invert` frequently.  Anyway, I believe exposing `aux_labels_sorted` at the global interface level will be confusing as users would never hold an FSA sorted by `aux_labels`.  Actually, your code don't hold such an FSA either, you just use `arc_aux_labels_sorted` as a local cache of the orginal FSA's `arc_sorted`.

Reply:
`arc_sorted` and `aux_labels_sorted `  are superseded by `properties`.